python/mlx/nn/layers/recurrent.py

```diff
@@ -0,0 +1,262 @@
+
+            n_x = x_proj_n[..., idx, :]
+
+            n_h = curr_hidden @ self.Whn
+            if self.bias:
+                n_h += self.bhn
+
+            n = n_x + r * n_h
+            n = mx.tanh(n)
+
+            curr_hidden = (1 - z) * n + z * curr_hidden
+            all_hidden.append(curr_hidden)
+
+        return mx.stack(all_hidden, axis=-2)
+
+
+class LSTM(Module):
+    r"""Apply one LSTM recurrent unit to a sequence of
+    shape ``NLD`` or ``LD``, where:
+        - ``N`` is the optional batch dimension
+        - ``L`` is the sequence length
+        - ``D`` is the input's feature dimension
+
+    Concretely, for each element of the sequence, computes:
+
+    .. math::
+        i_t = \sigma (W_{xi}x_t + b_{xi} + W_{hi}h_t + b_{hi})
+        f_t = \sigma (W_{xf}x_t + b_{xf} + W_{hf}h_t + b_{hf})
+        g_t = \text{tanh} (W_{xg}x_t + b_{xg} + W_{hg}h_t + b_{hg})
+        o_t = \sigma (W_{xo}x_t + b_{xo} + W_{ho}h_t + b_{ho})
+        c_{t + 1} = f_t \odot c_t + i_t \odot g_t
+        h_{t + 1} = o_t \text{tanh}(c_{t + 1})
+
+    The hidden state :math:`h` and cell state :math:`c` have shape ``NH``
+    or ``H``, depending whether the input is batched or not. Returns two
+    arrays: the hidden state and the cell state at each time step, both
+    of shape ``NLH`` or ``LH``.
+
+    Args:
+        input_size (int): Dimension of the input :math:`x`
+        hidden_size (int): Dimension of the hidden state :math:`h`
+        bias (bool): Whether to use biases or not. Default: `True`.
+    """
+
+    def __init__(
+        self,
+        input_size: int,
+        hidden_size: int,
+        bias: bool = True,
+    ):
+        super().__init__()
+        self.bias = bias
+        # i
+        self.Wxi = _weight_init(input_size, hidden_size)
+        self.Whi = _weight_init(hidden_size, hidden_size)
+        self.bi = _bias_init(hidden_size, bias)
+        # f
+        self.Wxf = _weight_init(input_size, hidden_size)
+        self.Whf = _weight_init(hidden_size, hidden_size)
+        self.bf = _bias_init(hidden_size, bias)
+        # g
+        self.Wxg = _weight_init(input_size, hidden_size)
+        self.Whg = _weight_init(hidden_size, hidden_size)
+        self.bg = _bias_init(hidden_size, bias)
+        # o
+        self.Wxo = _weight_init(input_size, hidden_size)
+        self.Who = _weight_init(hidden_size, hidden_size)
+        self.bo = _bias_init(hidden_size, bias)
+
+    def _extra_repr(self):
+        return f"input_dims={self.Wxi.shape[0]}, hidden_size={self.Whi.shape[0]}, bias={self.bias}"
+
+    def __call__(self, x):
+        x_proj_i = x @ self.Wxi
+        x_proj_f = x @ self.Wxf
+        x_proj_g = x @ self.Wxg
+        x_proj_o = x @ self.Wxo
+
+        if self.bias:
+            x_proj_i += self.bi
+            x_proj_f += self.bf
+            x_proj_g += self.bg
+            x_proj_o += self.bo
+
+        all_hidden = []
+        all_cell = []
+        curr_hidden = mx.zeros(shape=(self.Whi.shape[0],))
```

Comment:
Can we pass in `curr_hidden` and `curr_cell` as args, and default to zeros if they're `None`?

Reply:
Thank you for the comment, that's a good idea indeed; will update the code